### PR TITLE
Automating Results

### LINKED ISSUE #44: Automating Results
### **Feature request**:
Results are another main part of any machine learning project. Results consist of accuracy-loss plots, confusion matrix,  ROC curves, etc. Flow2Ml focuses on making the entire pipeline of machine learning much easier.  Thus In this feature issue, I propose another idea of automating the results.

### **The Idea**:
A class should be created which contains various methods for various functionalities by taking the trained models as input. With the trained models, the methods should generate various result metrics such as confession matrix, roc curves, etc. With all available result metrics finally, a report.pdf file should be created describing the entire process involved in the workflow.

### **Uses**:
With only a few lines of code, the entire results part could be completed and finally, a report would be generated which can be beneficial to the researchers.

<hr/>

**Side Note:**
If anyone has anything to add to this idea they are free to mention in this thread

### PR DESCRIPTION
**Pull request by ArkaprabhaChakraborty**
# Description:
A class has been created which contains various methods for various functionalities by taking the trained models as input. With the trained models, the methods generate various result metrics such as confession matrix, roc curves, etc. With all available result metrics finally, a report.docx file is created describing the entire process involved in the workflow.

dependencies:
sklearn
matplotlib
python-docx

Fixes #44 

## Type of change:

- [ ] New feature (non-breaking change which adds functionality)

 
# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by chebroluharika** [comment]
@ArkaprabhaChakraborty will this python-docx works for any OS?

File: flow2ml/AutoResults.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import matplotlib.pyplot as plt
+
+from sklearn.metrics import plot_confusion_matrix
+from sklearn.metrics import plot_precision_recall_curve
+from docx import Document
+from docx.shared import Inches
+
+class Auto_Results:
+    def __init__(self,model,X_test,Y_test):
+        self.model = model
```
Comment (chebroluharika) [inline flow2ml/AutoResults.py:13]: Please add documentation for every function that you defined here, so that anyone who see your code will understand easily.

File: flow2ml/AutoResults.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        self.prediction = self.model.predict_proba(X_test)
+        self.y_test = Y_test
+    
+    def roc_curve(self,figure_name="roc.jpeg"):
+        i= self.prediction.shape
+        try:
+            if (i[1]>1):
+                self.prediction = self.prediction[:, 1]
+        except:
+            pass
```
Comment (chebroluharika) [inline flow2ml/AutoResults.py:24]: throw with appropriate message

**Review by chebroluharika — CHANGES_REQUESTED** [review]
Please attach sample pdf file that you generated in this PR and also include it in folder(that you should create with name results
and also place testing code in Readme too
Can you make sure that this code works for image data also?

File: flow2ml/AutoResults.py (hunk 1 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml/AutoResults.py:13]: Okay

**Comment by ArkaprabhaChakraborty** [comment]
I have to check whether this works for image data or not, I tested it against the LogisticRegression model and it works pretty fine.

File: flow2ml/AutoResults.py (hunk 2 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml/AutoResults.py:24]: Okay.

**Comment by ArkaprabhaChakraborty** [comment]
Regarding python-docx, Its working is independent of any OS(or that's what its documentation says).

**Comment by yvkrishna** [comment]
@ArkaprabhaChakraborty, can you create a new file named "generate_pdf.py" and create a new class and paste the code for generating pdf into that file and inherit it in the AutoResults.py file. 

It would be better if all the code related to generating a pdf is available in a single file.

**Comment by ArkaprabhaChakraborty** [comment]
Okay Roger That!!!!!!!!!

**Comment by ArkaprabhaChakraborty** [comment]
Generating PDF actually involves the usage of a library called fpdf. Now creating a pdf file with that library is problematic so I thought of creating a doc file with python-docx and then converting it to pdf using the docx2pdf module. This is a simpler approach. Now should I create a generate_pdf.py file?

**Comment by ArkaprabhaChakraborty** [comment]
Maybe I'll change from the docx2pdf module because it has problems too. So I'll definitely keep you updated.

**Comment by yvkrishna** [comment]
> Generating PDF actually involves the usage of a library called fpdf. Now creating a pdf file with that library is problematic so I thought of creating a doc file with python-docx and then converting it to pdf using the docx2pdf module. This is a simpler approach. Now should I create a generate_pdf.py file?

can you checkout reportlab library

**Comment by yvkrishna** [comment]
or else you can send a pr for pdf generation later. and a new pr for only the results part

**Comment by ArkaprabhaChakraborty** [comment]
I have added the docstrings.

**Comment by ArkaprabhaChakraborty** [comment]
Now, where do I place the testing codes? I mean there's no test dir so should I make one and add the test.py?

**Comment by chebroluharika** [comment]
> Now, where do I place the testing codes? I mean there's no test dir so should I make one and add the test.py?

Code which are kept in bottom of generatepdf function, place the same in readme present inside flow2ml folder

**Comment by ArkaprabhaChakraborty** [comment]
Okay Got it.

**Comment by chebroluharika** [comment]
Please add here too 
https://github.com/flow2ml/Flow2ML/blob/52b8d5d9d604e91d82d2da9a8dee6e86dafcfd90/flow2ml/README.md

**Comment by ArkaprabhaChakraborty** [comment]
Added it there too.

File: flow2ml/AutoResults.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ def precision_recall_curve(self,figure_name="prc.jpeg"):
         plt.title('Precision Recall Curve', size = 35)
         plt.savefig(self.results_path+'/'+figure_name)
 
-    def get_results_pdf(self,file_name="report"):
+    def get_results_docx(self,file_name="report"):
```
Comment (chebroluharika) [inline flow2ml/AutoResults.py:67]: Will this docx display uniformly for any version of Windows? Because structure of docx changes according to Windows version. So please try to create pdf to store all images of results.

File: flow2ml/README.md [hunk 4]
```diff
@@ -14,6 +14,7 @@ Flow2Ml
         ├──Data_Loader.py
         ├──Filters.py  
         ├──Data_Augumentation.py
```
Comment (chebroluharika) [inline flow2ml/README.md:16]: Add usage part also here

File: flow2ml/AutoResults.py (hunk 3 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml/AutoResults.py:67]: Yes this docx will display uniformly for any version of windows. My later PR would involve converting this docx to pdf as there is no efficient way to do so yet.

File: flow2ml/README.md (hunk 4 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml/README.md:16]: Usage??? You mean test code?

Comment (chebroluharika) [reply flow2ml/README.md:16]: If you see in readme, there is a description of all functions used in files. Please add accordingly.

Comment (ArkaprabhaChakraborty) [reply flow2ml/README.md:16]: I mean In order to test this out... I have to add a model, test data and all. So  it will be better if I doit in a test directory

Comment (chebroluharika) [reply flow2ml/README.md:16]: I am not saying you to write code in readme. Just write description about what your functions do in autoresults.py. 
please check readme present in flow2ml folder to get idea

**Comment by ArkaprabhaChakraborty** [comment]
I have added a tests folder. Where I have shown a sample execution. Hope it helps

**Comment by ArkaprabhaChakraborty** [comment]
README.md has been updated too

File: flow2ml.egg-info/SOURCES.txt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 README.md
 setup.py
+flow2ml/Auto_Results.py
```
Comment (chebroluharika) [inline flow2ml.egg-info/SOURCES.txt:3]: Please remove this. All this content gets generated when we publish package to Pypi

File: flow2ml/__init__.py [hunk 6]
```diff
@@ -1 +1,2 @@
-from .Flow import Flow
\ No newline at end of file
+from .Flow import Flow
+from .Auto_Results import Auto_Results
```
Comment (chebroluharika) [inline flow2ml/__init__.py:2]: Can you tell me the reason for adding this?

File: tests/test.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
```
Comment (chebroluharika) [inline tests/test.py:1]: In tests folder, we generally have unit tests in it. So please remove here and place all the code in Readme which is in flow2ml folder(not in global Readme) itself

File: tests/winequality.names [hunk 8]
```diff
@@ -0,0 +1,72 @@
+Citation Request:
```
Comment (chebroluharika) [inline tests/winequality.names:1]: Please remove this file

**Comment by chebroluharika** [comment]
@ArkaprabhaChakraborty please make sure this code works for images dataset, because this package solely meant for image dataset as of now

File: flow2ml/__init__.py (hunk 6 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml/__init__.py:2]: Otherwise ... I cant Import it from flow2ml package.

File: flow2ml.egg-info/SOURCES.txt (hunk 5 above)
Comment (ArkaprabhaChakraborty) [reply flow2ml.egg-info/SOURCES.txt]: I build the project on my local pc to test things out. So the sources.txt got updated automatically.

**Comment by ArkaprabhaChakraborty** [comment]
It works for images, non images too.

File: tests/winequality.names (hunk 8 above)
Comment (ArkaprabhaChakraborty) [reply tests/winequality.names]: Removed

**Comment by ArkaprabhaChakraborty** [comment]
About the docx file generated by the code on a test model.
[report.docx](https://github.com/flow2ml/Flow2ML/files/6436714/report.docx)
